src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,389 @@
+                }
+                _ => continue,
+            }
+        }
+
+        if !data_slice_batch_info.data_slices_batch.is_empty() {
+            data_slice_batches.push(data_slice_batch_info);
+        }
+
+        Ok(ChainFilesInfo {
+            data_slice_batches,
+            _block_file_urls: block_file_urls,
+            old_head: old_head.unwrap(),
+        })
+    }
+
+    async fn merge_files(
+        &self,
+        data_slice_batches: &mut Vec<DataSliceBatchInfo>,
+        dataset_name: &DatasetName,
+        compact_dir_path: &Path,
+    ) -> Result<(), CompactError> {
+        let ctx = SessionContext::new();
+
+        for data_slice_batch_info in data_slice_batches {
+            if data_slice_batch_info.data_slices_batch.len() == 1 {
+                continue;
+            }
+            let data_frame = ctx
+                .read_parquet(
+                    data_slice_batch_info.data_slices_batch.clone(),
+                    datafusion::execution::options::ParquetReadOptions {
+                        file_extension: "",
+                        ..Default::default()
+                    },
+                )
+                .await
+                .int_err()?;
+
+            data_slice_batch_info.new_file_path = Some(compact_dir_path.join(
+                get_random_operation_name_with_prefix(format!("{dataset_name}-").as_str()),
+            ));
+
+            data_frame
+                .write_parquet(
+                    data_slice_batch_info
+                        .new_file_path
+                        .as_ref()
+                        .unwrap()
+                        .to_str()
+                        .unwrap(),
+                    datafusion::dataframe::DataFrameWriteOptions::new()
+                        .with_single_file_output(true),
+                    None,
+                )
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+
+    fn create_run_compact_dir(&self) -> Result<PathBuf, CompactError> {
+        let compact_dir_path = self.run_info_dir.join("compact");
+        fs::create_dir_all(&compact_dir_path).int_err()?;
+        Ok(compact_dir_path)
+    }
+
+    async fn commit_new_blocks(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        chain_files_info: &ChainFilesInfo,
+    ) -> Result<(Vec<Url>, Multihash), CompactError> {
+        let chain = dataset.as_metadata_chain();
+
+        let mut current_head = chain_files_info.old_head.clone();
+        let mut old_data_slices: Vec<Url> = vec![];
+
+        for data_slice_batch_info in chain_files_info.data_slice_batches.iter().rev() {
+            if let Some(block_hash) = data_slice_batch_info.hash.as_ref() {
+                let block = chain.get_block(block_hash).await.int_err()?;
+
+                if let MetadataEvent::AddData(add_data_event) = block.event {
+                    let block_event = MetadataEvent::AddData(AddData {
+                        prev_checkpoint: None,
```

Comment:
Ingest engines may actually produce checkpoints, so we need to preserve them. When merging two `AddData` blocks the resulting block's checkpoint should be one of the last block.

Reply:
Added checkpoint handling